diff --git a/kernel/module.c b/kernel/module.c
index 529efae9f481..2af3d03d114b 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -61,6 +61,7 @@
 #include <linux/pfn.h>
 #include <linux/bsearch.h>
 #include <linux/dynamic_debug.h>
+#include <linux/delay.h>
 #include <uapi/linux/module.h>
 #include "module-internal.h"
 
@@ -1762,6 +1763,12 @@ out:
 	return err;
 }
 
+static void mod_hack(struct module *mod)
+{
+	if (strcmp(mod->name, "xfs") == 0)
+		msleep(10);
+}
+
 static int mod_sysfs_setup(struct module *mod,
 			   const struct load_info *info,
 			   struct kernel_param *kparam,
@@ -1783,6 +1790,8 @@ static int mod_sysfs_setup(struct module *mod,
 	if (err)
 		goto out_unreg_holders;
 
+	mod_hack(mod);
+
 	err = module_add_modinfo_attrs(mod);
 	if (err)
 		goto out_unreg_param;
